File: bindings/go/peer.go
package blazen

import (
	"context"
	"encoding/json"
	"fmt"
	"runtime"
	"sync"
	"time"

	uniffiblazen "github.com/zachhandley/Blazen/bindings/go/internal/uniffi/blazen"
)

// PeerClient is a client handle for invoking workflows on a remote
// [PeerServer] over gRPC. RPCs go out over a multiplexed HTTP/2 channel
// held inside the client; multiple concurrent calls on the same
// PeerClient are safe and share the connection.
//
// Construct one via [ConnectPeerClient]. Call [PeerClient.Close] when
// finished to release the underlying native handle; a finalizer is
// attached as a safety net, but explicit Close is preferred for
// predictable resource release.
type PeerClient struct {
	inner *uniffiblazen.PeerClient
	once  sync.Once
}

// ConnectPeerClient establishes a connection to a remote Blazen peer.
//
// address is the peer's gRPC endpoint URI (for example
// "http://10.0.0.1:50051" or "https://node-a.local:7443").
// clientNodeID is a local identifier stamped into outgoing requests for
// trace correlation on both sides of the connection; the local hostname
// or a process-startup UUID work well.
//
// This call blocks while the TCP/HTTP-2 handshake completes — the
// upstream FFI constructor drives the connect on the shared Tokio
// runtime so synchronous Go code can set up a client without an async
// story. A [PeerError] with Kind "Transport" is returned if the URI is
// invalid or the handshake fails.
func ConnectPeerClient(address, clientNodeID string) (*PeerClient, error) {
	ensureInit()
	inner, err := uniffiblazen.PeerClientConnect(address, clientNodeID)
	if err != nil {
		return nil, wrapErr(err)
	}
	c := &PeerClient{inner: inner}
	// Safety net: if the caller forgets Close, the finalizer reclaims the
	// underlying native handle on GC. Close() clears the finalizer via
	// the underlying Destroy().
	runtime.SetFinalizer(c, func(pc *PeerClient) { pc.Close() })
	return c, nil
}

// NodeID returns the node-id this client stamps into outgoing requests
// for trace correlation. Returns the empty string after [PeerClient.Close].
func (c *PeerClient) NodeID() string {
	if c == nil || c.inner == nil {
		return ""
	}
	return c.inner.NodeId()
}

// RunRemoteWorkflow invokes a workflow on the remote peer and waits for
// its terminal result.
//
// workflowName is the symbolic name the remote peer's step registry
// knows the workflow as. stepIDs lists the step identifiers to execute
// in order; every entry must be registered on the remote peer's process
// or the call fails with a [PeerError] of Kind "UnknownStep". input is
// JSON-marshalled into the workflow's entry payload — passing nil sends
// the JSON literal `null`. timeout bounds the remote workflow's
// wall-clock execution; pass 0 to defer to the peer's default deadline.
// Sub-second durations are rounded up to the next whole second.
//
// The Rust side of the call is blocking from Go's perspective. To
// honour ctx cancellation the FFI call runs on a background goroutine
// and the function selects on ctx.Done(); when ctx fires first the
// function returns ctx.Err() immediately. The remote workflow itself
// keeps running until it finishes naturally — cancellation propagation
// across the wire is a known gap pending an upstream UniFFI feature.
//
// The returned [WorkflowResult] carries the terminal payload
// synthesised from the remote response. Per-run LLM token usage and
// cost are not propagated over the wire and the returned result reports
// them as zero; callers needing those should query the remote peer's
// telemetry directly.
func (c *PeerClient) RunRemoteWorkflow(ctx context.Context, workflowName string, stepIDs []string, input any, timeout time.Duration) (*WorkflowResult, error) {
	payload, err := json.Marshal(input)
	if err != nil {
		return nil, &ValidationError{Message: fmt.Sprintf("marshal remote workflow input: %s", err)}
	}
	return c.RunRemoteWorkflowJSON(ctx, workflowName, stepIDs, string(payload), timeout)
}

// RunRemoteWorkflowJSON is the lower-level form of
// [PeerClient.RunRemoteWorkflow] taking a pre-JSON payload string. Use
// this when the input is already a JSON document and the extra
// marshal/unmarshal of RunRemoteWorkflow would be wasteful.
//
// ctx cancellation, timeout, and error semantics match
// [PeerClient.RunRemoteWorkflow].
func (c *PeerClient) RunRemoteWorkflowJSON(ctx context.Context, workflowName string, stepIDs []string, inputJSON string, timeout time.Duration) (*WorkflowResult, error) {
	if c == nil || c.inner == nil {
		return nil, &ValidationError{Message: "peer client has been closed"}
	}
	timeoutSecs := durationToSecondsPtr(timeout)

	type runResult struct {
		res uniffiblazen.WorkflowResult
		err error
	}
	done := make(chan runResult, 1)
	go func() {
		res, err := c.inner.RunRemoteWorkflow(workflowName, stepIDs, inputJSON, timeoutSecs)
		done <- runResult{res: res, err: err}
	}()

	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case r := <-done:
		if r.err != nil {
			return nil, wrapErr(r.err)
		}
		out := workflowResultFromFFI(r.res)
		return &out, nil
	}
}

// Close releases the underlying native handle, closing the gRPC
// connection. It is safe to call Close multiple times and from multiple
// goroutines; subsequent calls are no-ops.
func (c *PeerClient) Close() {
	c.once.Do(func() {
		if c.inner != nil {
			c.inner.Destroy()
			c.inner = nil
		}
	})
}

// PeerServer is a node-local Blazen peer gRPC server that accepts
// inbound workflow requests from remote peers.
//
// Construct one with [NewPeerServer] and start the listener with
// [PeerServer.Serve] or [PeerServer.ServeBlocking]. Dispatched
// workflows are resolved at request time through the process-wide step
// registry, so any workflow whose steps have been registered in this
// process can be invoked remotely by name.
//
// Call [PeerServer.Close] when finished to release the underlying
// native handle; a finalizer is attached as a safety net, but explicit
// Close is preferred for predictable resource release.
type PeerServer struct {
	inner *uniffiblazen.PeerServer
	once  sync.Once
}

// NewPeerServer constructs an inbound peer server identifiable as
// nodeID. The nodeID is the stable identifier this server stamps onto
// every remote-ref descriptor it returns; the hostname or a UUID picked
// at process startup work well.
//
// The server is not listening yet — call [PeerServer.Serve] or
// [PeerServer.ServeBlocking] to bind. The constructor itself is
// infallible.
func NewPeerServer(nodeID string) *PeerServer {
	ensureInit()
	s := &PeerServer{inner: uniffiblazen.NewPeerServer(nodeID)}
	// Safety net: if the caller forgets Close, the finalizer reclaims the
	// underlying native handle on GC. Close() clears the finalizer via
	// the underlying Destroy().
	runtime.SetFinalizer(s, func(ps *PeerServer) { ps.Close() })
	return s
}

// Serve binds the gRPC server to listenAddress (for example ":50051" or
// "0.0.0.0:50051") and blocks serving requests until the underlying
// socket fails or the server is consumed.
//
// listenAddress must parse as a standard socket address. The underlying
// server is consumed by this call; invoking Serve or ServeBlocking a
// second time on the same PeerServer returns a [ValidationError]. A
// [PeerError] with Kind "Transport" is returned if the listener fails
// to bind or hits a fatal I/O error while serving.
//
// The Rust side of the call is blocking from Go's perspective. To
// honour ctx cancellation the FFI call runs on a background goroutine
// and the function selects on ctx.Done(); on cancellation the wait
// unblocks immediately, but the Rust-side server keeps running until it
// finishes naturally (cancellation propagation into the runtime is a
// known gap pending an upstream UniFFI feature).
func (s *PeerServer) Serve(ctx context.Context, listenAddress string) error {
	if s == nil || s.inner == nil {
		return &ValidationError{Message: "peer server has been closed"}
	}
	done := make(chan error, 1)
	go func() {
		done <- s.inner.Serve(listenAddress)
	}()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case err := <-done:
		if err != nil {
			return wrapErr(err)
		}
		return nil
	}
}

// ServeBlocking is the synchronous form of [PeerServer.Serve]. It
// blocks the calling goroutine until the server exits.
//
// Error semantics match [PeerServer.Serve].
func (s *PeerServer) ServeBlocking(listenAddress string) error {
	if s == nil || s.inner == nil {
		return &ValidationError{Message: "peer server has been closed"}
	}
	if err := s.inner.ServeBlocking(listenAddress); err != nil {
		return wrapErr(err)
	}
	return nil
}

// Close releases the underlying native handle. It is safe to call Close
// multiple times and from multiple goroutines; subsequent calls are
// no-ops.
func (s *PeerServer) Close() {
	s.once.Do(func() {
		if s.inner != nil {
			s.inner.Destroy()
			s.inner = nil
		}
	})
}

// durationToSecondsPtr converts a Go time.Duration into the optional
// uint64 second count expected by the FFI's timeout parameter. A
// non-positive duration maps to nil (deferring to the remote peer's
// default deadline); sub-second positive durations round up so a
// caller-supplied 500ms timeout still allows the remote a full second.
func durationToSecondsPtr(d time.Duration) *uint64 {
	if d <= 0 {
		return nil
	}
	secs := uint64((d + time.Second - 1) / time.Second)
	return &secs
}

File: crates/blazen-uniffi/src/peer.rs
//! Distributed peer surface for the UniFFI bindings.
//!
//! Wraps upstream [`blazen_peer::BlazenPeerServer`] and
//! [`blazen_peer::BlazenPeerClient`] so foreign callers can run Blazen
//! workflows across processes / machines via the peer gRPC transport.
//!
//! ## Deviations from `blazen-py`
//!
//! `blazen-py` mirrors the upstream `BlazenPeerClient::invoke_sub_workflow`
//! surface directly, including the full `SubWorkflowRequest` /
//! `SubWorkflowResponse` envelope (with state map and `remote_refs`). UniFFI
//! does not handle nested record graphs across four foreign runtimes as
//! cleanly as PyO3, so this layer flattens the response down to the bits a
//! workflow caller actually consumes — the terminal `StopEvent` payload —
//! and exposes it as the same [`crate::workflow::WorkflowResult`] type the
//! in-process [`crate::workflow::Workflow::run`] returns. Session-ref
//! proxies (`remote_refs`, `deref_session_ref`, `release_session_ref`) and
//! the streaming-events RPC are deferred to a follow-up; today's surface
//! covers fire-and-forget remote execution with a JSON result.
//!
//! ## Deviations from upstream
//!
//! Upstream [`blazen_peer::BlazenPeerServer`] does not have a
//! `register_workflow` API or a `start` / `stop` lifecycle — it dispatches
//! every invocation by step ID through the process-wide
//! [`blazen_core::step_registry`] and its `serve` method takes `self` by
//! value and runs until the bound socket errors. The originally-requested
//! `register_workflow` / `start` / `stop` shape therefore cannot be honoured
//! without inventing a parallel workflow catalog in this crate; instead,
//! [`PeerServer::serve`] (and its blocking sibling) is the single entry
//! point, matching upstream and matching the Python binding's
//! [`PyBlazenPeerServer::serve`] method.

use std::net::SocketAddr;
use std::sync::Arc;

use blazen_peer::{BlazenPeerClient, BlazenPeerServer, SubWorkflowRequest};
use tokio::sync::Mutex;

use crate::errors::{BlazenError, BlazenResult};
use crate::runtime::runtime;
use crate::workflow::{Event, WorkflowResult};

/// Node-local Blazen peer gRPC server.
///
/// Owns a stable `node_id` embedded in every
/// `RemoteRefDescriptor` this peer hands out, plus an in-process session-ref
/// registry. Construct with [`PeerServer::new`] and start the gRPC listener
/// with [`PeerServer::serve`] (async) or [`PeerServer::serve_blocking`].
///
/// Dispatched workflows are resolved at request time through the
/// process-wide [`blazen_core::step_registry`], so any workflow whose steps
/// have been registered in this process can be invoked remotely by name.
#[derive(uniffi::Object)]
pub struct PeerServer {
    inner: Mutex<Option<BlazenPeerServer>>,
}

#[uniffi::export]
impl PeerServer {
    /// Create a new peer server with a fresh, empty session-ref registry.
    ///
    /// `node_id` is the stable identifier that this server stamps onto every
    /// `RemoteRefDescriptor` it returns. Typical values are the hostname or
    /// a UUID picked at process startup.
    #[uniffi::constructor]
    #[must_use]
    pub fn new(node_id: String) -> Arc<Self> {
        Arc::new(Self {
            inner: Mutex::new(Some(BlazenPeerServer::new(node_id))),
        })
    }
}

#[uniffi::export(async_runtime = "tokio")]
impl PeerServer {
    /// Bind the gRPC server to `listen_address` and serve until the
    /// listener errors or the caller's async task is cancelled.
    ///
    /// `listen_address` must parse as a [`std::net::SocketAddr`] (for
    /// example `"0.0.0.0:50051"` or `"127.0.0.1:7443"`). This method
    /// consumes the underlying server; calling it twice on the same
    /// `PeerServer` returns [`BlazenError::Validation`].
    ///
    /// # Errors
    ///
    /// Returns [`BlazenError::Validation`] if `listen_address` cannot be
    /// parsed or the server has already been consumed by a prior call, and
    /// [`BlazenError::Peer`] (`kind = "Transport"`) if the gRPC listener
    /// fails to bind or encounters a fatal I/O error while serving.
    pub async fn serve(self: Arc<Self>, listen_address: String) -> BlazenResult<()> {
        let addr: SocketAddr = listen_address
            .parse()
            .map_err(|e| BlazenError::Validation {
                message: format!("invalid peer listen address {listen_address:?}: {e}"),
            })?;
        let server = {
            let mut guard = self.inner.lock().await;
            guard.take().ok_or(BlazenError::Validation {
                message: "PeerServer already consumed by a prior serve() call".into(),
            })?
        };
        server.serve(addr).await.map_err(BlazenError::from)
    }
}

#[uniffi::export]
impl PeerServer {
    /// Synchronous variant of [`PeerServer::serve`] — blocks the current
    /// thread on the shared Tokio runtime until the server exits. Intended
    /// for foreign callers (Ruby scripts, Go `main`, Swift CLIs) that want
    /// a one-shot blocking bind without driving an async runtime.
    ///
    /// # Errors
    ///
    /// Same as [`PeerServer::serve`].
    pub fn serve_blocking(self: Arc<Self>, listen_address: String) -> BlazenResult<()> {
        let this = Arc::clone(&self);
        runtime().block_on(async move { this.serve(listen_address).await })
    }
}

/// Client handle for invoking workflows on a remote [`PeerServer`].
///
/// Construct with [`PeerClient::connect`]. RPCs go out over a multiplexed
/// HTTP/2 channel held inside the client; multiple concurrent calls on the
/// same `PeerClient` are safe and share the connection.
#[derive(uniffi::Object)]
pub struct PeerClient {
    inner: Mutex<BlazenPeerClient>,
    node_id: String,
}

#[uniffi::export]
impl PeerClient {
    /// Open a connection to the peer at `address`.
    ///
    /// `address` must be a valid gRPC endpoint URI such as
    /// `"http://node-a.local:7443"`. `client_node_id` identifies *this* end
    /// of the connection in trace logs on both sides and is typically the
    /// local hostname or a process-startup UUID.
    ///
    /// This constructor is blocking — it drives the TCP connect on the
    /// shared Tokio runtime so foreign callers without an async story
    /// (Ruby, synchronous Go code) can still set up a client. The async
    /// connect path is internal to upstream `BlazenPeerClient` and is not
    /// re-exposed across UniFFI to avoid a constructor that returns a
    /// coroutine in every target language.
    ///
    /// # Errors
    ///
    /// Returns [`BlazenError::Peer`] (`kind = "Transport"`) if the
    /// endpoint URI is invalid or the TCP / HTTP/2 handshake fails.
    #[uniffi::constructor]
    pub fn connect(address: String, client_node_id: String) -> BlazenResult<Arc<Self>> {
        let node_id_for_client = client_node_id.clone();
        let client = runtime()
            .block_on(async move { BlazenPeerClient::connect(address, node_id_for_client).await })
            .map_err(BlazenError::from)?;
        Ok(Arc::new(Self {
            inner: Mutex::new(client),
            node_id: client_node_id,
        }))
    }

    /// The node-id stamped into outgoing requests for tracing.
    #[must_use]
    pub fn node_id(self: Arc<Self>) -> String {
        self.node_id.clone()
    }
}

#[uniffi::export(async_runtime = "tokio")]
impl PeerClient {
    /// Invoke a workflow on the connected peer and wait for its terminal
    /// result.
    ///
    /// - `workflow_name` is the symbolic name the remote peer's
    ///   [`blazen_core::step_registry`] knows the workflow as.
    /// - `step_ids` is the ordered list of step identifiers to execute.
    ///   Every entry must be registered on the remote peer's process or
    ///   the call fails with [`BlazenError::Peer`] (`kind = "UnknownStep"`).
    ///   This is required by the peer wire protocol — see
    ///   [`blazen_peer::SubWorkflowRequest`].
    /// - `input_json` is the JSON-encoded payload fed into the workflow's
    ///   entry step.
    /// - `timeout_secs` bounds the remote workflow's wall-clock execution.
    ///   `None` defers to the server's default deadline.
    ///
    /// The returned [`WorkflowResult`] carries the terminal `StopEvent`
    /// payload synthesised from the remote `SubWorkflowResponse`. Per-run
    /// LLM token usage and cost are not propagated over the wire and are
    /// reported as zero; foreign callers needing those should query the
    /// remote peer's telemetry directly.
    ///
    /// # Errors
    ///
    /// Returns [`BlazenError::Peer`] for encode / transport / envelope-
    /// version failures, or [`BlazenError::Workflow`] when the remote
    /// reports a workflow-execution error in `SubWorkflowResponse::error`.
    pub async fn run_remote_workflow(
        self: Arc<Self>,
        workflow_name: String,
        step_ids: Vec<String>,
        input_json: String,
        timeout_secs: Option<u64>,
    ) -> BlazenResult<WorkflowResult> {
        let input: serde_json::Value = serde_json::from_str(&input_json)?;
        let request = SubWorkflowRequest::new(workflow_name, step_ids, &input, timeout_secs)
            .map_err(BlazenError::from)?;
        let response = {
            let mut guard = self.inner.lock().await;
            guard
                .invoke_sub_workflow(request)
                .await
                .map_err(BlazenError::from)?
        };
        sub_workflow_response_to_result(response)
    }
}

#[uniffi::export]
impl PeerClient {
    /// Synchronous variant of [`PeerClient::run_remote_workflow`] — blocks
    /// the current thread on the shared Tokio runtime.
    ///
    /// # Errors
    ///
    /// Same as [`PeerClient::run_remote_workflow`].
    pub fn run_remote_workflow_blocking(
        self: Arc<Self>,
        workflow_name: String,
        step_ids: Vec<String>,
        input_json: String,
        timeout_secs: Option<u64>,
    ) -> BlazenResult<WorkflowResult> {
        let this = Arc::clone(&self);
        runtime().block_on(async move {
            this.run_remote_workflow(workflow_name, step_ids, input_json, timeout_secs)
                .await
        })
    }
}

/// Convert an upstream [`blazen_peer::SubWorkflowResponse`] into the
/// flattened [`WorkflowResult`] surface UniFFI exposes.
///
/// The remote `result_json` field (the terminal `StopEvent` payload) becomes
/// the `data_json` of a synthesised `StopEvent`-typed [`Event`]; an empty
/// remote result is mapped to the JSON literal `null`. Remote token usage
/// and cost are not carried in the peer wire format and default to zero.
fn sub_workflow_response_to_result(
    response: blazen_peer::SubWorkflowResponse,
) -> BlazenResult<WorkflowResult> {
    if let Some(error) = response.error {
        return Err(BlazenError::Workflow { message: error });
    }
    let result_value = response.result_value().map_err(BlazenError::from)?;
    let data_json = match result_value {
        Some(value) => value.to_string(),
        None => "null".to_string(),
    };
    Ok(WorkflowResult {
        event: Event {
            event_type: "StopEvent".to_string(),
            data_json,
        },
        total_input_tokens: 0,
        total_output_tokens: 0,
        total_cost_usd: 0.0,
    })
}
